src/wallet/wallet.cpp

```diff
@@ -595,10 +595,16 @@ void CWallet::ChainTipAdded(const CBlockIndex *pindex,
         // Don't flush during startup.
         nLastSetChain = nNow;
     }
-    if (nLastSetChain + (int64_t)DATABASE_WRITE_INTERVAL * 1000000 < nNow) {
+    if (nLastSetChain + (int64_t)WITNESS_WRITE_INTERVAL * 1000000 < nNow) {
         nLastSetChain = nNow;
-        LOCK(cs_main);
-        SetBestChain(chainActive.GetLocator());
+        CBlockLocator loc;
+        {
+            // The locator must be derived from the pindex used to increment
+            // the witnesses above; pindex can be behind chainActive.Tip().
+            LOCK(cs_main);
+            loc = chainActive.GetLocator(pindex);
+        }
+        SetBestChain(loc);
```

Comment:
The `cs_main` lock shouldn't be held during `SetBestChain()` because it synchronously writes to disk.